clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java

```diff
@@ -129,21 +130,29 @@ public void clearAssignment() {
     }
 
     public Long fetched(TopicPartition tp) {
-        return this.fetched.get(tp);
+        return assertAssignment(tp).fetched;
     }
 
     public void fetched(TopicPartition tp, long offset) {
-        if (!this.assignedPartitions.contains(tp))
-            throw new IllegalArgumentException("Can't change the fetch position for a partition you are not currently subscribed to.");
-        this.fetched.put(tp, offset);
+        if (!isFetchable(tp))
+            // we only allow updates to the fetched position when both fetched and consumed positions are valid
+            throw new IllegalStateException("Cannot update fetch position of a partition not in a fetchable state ");
+        assertAssignment(tp).fetched(offset);
+    }
+
+    private TopicPartitionState assertAssignment(TopicPartition tp) {
```

Comment:
I added "assert" in the name to emphasize the expectation that the method will throw an error if the partition is not assigned. I'm not sure that would be as clear with either of these suggestions, but perhaps it's not that big of a deal. How about "assignedState"?